### PR TITLE
feat: use `nuxt-shiki`

### LINKED ISSUE #148: Use `nuxt-shiki` for syntax highlighting
Hi. Today I learned (https://github.com/unjs/undocs/issues/45) that there are several places Shiki is being integrated into Nuxt (MDC and UI at least I know) and also maintaining/well-testing shiki with an optimized integration is becoming trickier every day including tree-shaking, configuration, and wasm support PS if someone wants directly use shiki API for code highlighting, they need to again duplicate logics (like un undocs)

I made a quick draft of [`nuxt-shiki`](https://github.com/pi0/nuxt-shiki) module to make it more standard from one place.

I would love to hear what would be needed / what features are missing to make this happen.

(UI discussion https://github.com/nuxt/ui/issues/1361)

### PR DESCRIPTION
**Pull request by cpreston321**
resolves #148

In this PR I added the module `nuxt-shiki`. 

Which controls the main core level to shiki like loading wasm binary, exposes functions for other projects to then consume in there projects.

cc:// @Atinux  @pi0

### CODE REVIEW

File: src/module.ts [hunk 1]
```diff
@@ -63,26 +63,12 @@ export default defineNuxtModule<ModuleOptions>({
...
-          })
-        }
-        addWasmSupport(nitro)
-        nitro.hooks.hook('prerender:init', (prerenderer) => {
-          addWasmSupport(prerenderer)
-        })
+      // Install nuxt-shiki
+      // @see https://github.com/pi0/nuxt-shiki#nuxt-shiki
+      installModule('nuxt-shiki', {
+        bundledThemes: options.highlight.themes,
```
Comment (pi0) [inline src/module.ts:69]: Can you test if it still extends `shiki:` config key options? (normally kit should merge just double checking)

**Review by pi0 — APPROVED** [review]
Thanks. LGTM with overall changes.

Hnading over to @farnabaz to see what is needed.

File: src/module.ts (hunk 1 above)
Comment (cpreston321) [reply src/module.ts:69]: I can test this, the only reason I passed it in was because it was defined on the mdc config already.

Comment (cpreston321) [reply src/module.ts:69]: Just tested and it still can extend from `shiki:`

**Comment by cpreston321** [comment]
@pi0 Since the module is installed and `loadShiki` is under `#imports`, whats a good way to test this? I haven't found a good way to import without the ci:test throwing errors. 

CleanShot 2024-03-14 at 17 01 48@2x

**Comment by pi0** [comment]
Checking locally, we need to update `#mdc-highlighter` / `mdc-highlighter.js` template as well.

I will check tomorrow more too. (also /cc @antfu if you have any recommendations for migration)

**Comment by farnabaz** [comment]
Using `nuxt-shiki` will break the possibility of using `@nuxtjs/mdc` package outside Nuxt scope.

Currently, there are some use-cases that we are using MDC module to parse/render markdown files in non-Nuxt applications.

`@nuxtjs/mdc` is an entry point for users to use MDC syntax and should work outside Nuxt too.

**Comment by pi0** [comment]
Interesting point... I am open to also expose runtime of `nuxt-shiki` to be directly usable if that would help adoption.

**Comment by pi0** [comment]
on it

**Comment by farnabaz** [comment]
What do you think of improving the usage in MDC instead of separating the logic into another package?
For me, Shiki could be part of MDC module and users can use it separately if they need to.

**Comment by pi0** [comment]
- Maintaining shiki/wasm support within another module is super tricky today and we faced multiple regressions
- When users need to use Shiki outside of mdc, it will have become super tricky as well with multiple implementations 

I am building `nuxt-shiki` in good faith for our entire ecosystem and if you have any more needs that need to be considered 💯 welcome.

**Comment by farnabaz** [comment]
I understand how tricky it is. I'm not against building `nuxt-shiki`, IMO we can just take similar approach in MDC module to improve usability, unify the implementation.

I just don't get why it needs to be done in a separate module. What is the limitation that we need to separate it? Maybe I'm missing something.

**Comment by pi0** [comment]
Simply not duplicating the same thing and reducing maintenance/sync costs. Sent you a message in Slack if you like to talk :)

**Comment by farnabaz** [comment]
> * Maintaining shiki/wasm support within another module is super tricky today and we faced multiple regressions
> * When users need to use Shiki outside of mdc, it will have become super tricky as well with multiple implementations
> 
> I am building `nuxt-shiki` in good faith for our entire ecosystem and if you have any more needs that need to be considered 💯 welcome. (PS: you are already a maintainer of `nuxt-shiki` :)

Again I'm not against `nuxt-shiki` nor standing in front of it, I asked to improve it in MDC module. And so far I didn't get why we can't do all improvments inside MDC module and expose same/similar helpers.

Anyway like I said in https://github.com/nuxt/ui/issues/1361#issuecomment-2010376067 we can adopt it.
Feel free to update this PR and we can merge it  :)

**Comment by pi0** [comment]
@cpreston321 @farnabaz  I am no longer interested in this topic. feel free to continue on this PR or close...

**Comment by cpreston321** [comment]
I will close this PR.

**Comment by tobiasdiez** [comment]
This PR seemed to be triggered by https://github.com/unjs/undocs/issues/100
```
13:34:21.688 | [error] [nuxt] [request error] [unhandled] [500] Unknown file extension ".wasm" for /opt/buildhome/repo/docs/node_modules/shiki/dist/onig.wasm
```

I'm now experiencing the same error but with ui-pro (docs), see https://github.com/nuxt-modules/storybook/pull/690. What's the best/recommended way to fix such an error now?

**Pull request closed without merging**